ascent_macro: Fix minimal versions (by upgrading syn to v1.0.109, itertools to v0.12)

While a bump to >= v2.x would [obviously be preferable](https://github.com/s-arash/ascent/issues/18) I'd like to suggest bumping it at least to the highest v1.x version available for the time being.

Deleting `Cargo.lock` and running `cargo minimal-versions check --workspace --all-features --ignore-private -v` ([crates.io](https://crates.io/crates/cargo-minimal-versions)) reveals that `ascent_macro` is using APIs from `syn` and `itertools` that are not available in their versions specified in `Cargo.toml`

<details>

<summary>Full terminal log</summary>

```terminal
$ cargo minimal-versions check --workspace --all-features --ignore-private -v                                                                            
(base) 
info: modifying from <SNIP>/ascent/Cargo.toml
info: modifying from <SNIP>/ascent_base/Cargo.toml
info: modifying from <SNIP>/ascent_macro/Cargo.toml
info: running `rustup run nightly cargo update -Z minimal-versions`
warning: virtual workspace defaulting to `resolver = "1"` despite one or more workspace members being on edition 2021 which implies `resolver = "2"`
note: to keep the current resolver, specify `workspace.resolver = "1"` in the workspace root's manifest
note: to use the edition 2021 resolver, specify `workspace.resolver = "2"` in the workspace root's manifest
note: for more details see https://doc.rust-lang.org/cargo/reference/resolver.html#resolver-versions
    Updating crates.io index
info: running `~/.rustup/toolchains/stable-x86_64-apple-darwin/bin/cargo hack check --workspace --all-features`
info: running `cargo check --all-features` on ascent (1/3)
warning: virtual workspace defaulting to `resolver = "1"` despite one or more workspace members being on edition 2021 which implies `resolver = "2"`
note: to keep the current resolver, specify `workspace.resolver = "1"` in the workspace root's manifest
note: to use the edition 2021 resolver, specify `workspace.resolver = "2"` in the workspace root's manifest
note: for more details see https://doc.rust-lang.org/cargo/reference/resolver.html#resolver-versions
    Blocking waiting for file lock on package cache
    Blocking waiting for file lock on package cache
    Blocking waiting for file lock on package cache
   Compiling crossbeam-utils v0.8.7
   Compiling proc-macro2 v1.0.52
   Compiling memoffset v0.6.1
    Checking lazy_static v1.4.0
   Compiling unicode-ident v1.0.0
   Compiling libc v0.2.95
    Checking scopeguard v1.1.0
   Compiling crossbeam-epoch v0.9.5
   Compiling syn v1.0.0
   Compiling unicode-xid v0.2.0
   Compiling ahash v0.8.4
   Compiling indexmap v1.6.2
   Compiling paste v1.0.0
    Checking either v1.0.0
    Checking zerocopy v0.7.3
   Compiling hashbrown v0.9.1
    Checking smallvec v1.6.1
    Checking allocator-api2 v0.2.9
   Compiling fixedbitset v0.4.0
   Compiling heck v0.4.0
   Compiling itertools v0.10.0
    Checking lock_api v0.4.10
    Checking instant v0.1.0
    Checking segvec v0.1.0
    Checking sync-unsafe-cell v0.1.0
   Compiling ascent_base v0.5.0 (<SNIP>/ascent_base)
    Checking crossbeam-channel v0.5.0
   Compiling petgraph v0.6.0
    Checking crossbeam-deque v0.8.1
    Checking num_cpus v1.2.0
    Checking parking_lot_core v0.9.8
    Checking rayon-core v1.11.0
   Compiling quote v1.0.26
    Checking rayon v1.7.0
   Compiling proc-macro-error-attr v1.0.4
    Checking hashbrown v0.14.0
    Checking dashmap v5.5.0
   Compiling proc-macro-error v1.0.4
   Compiling derive-syn-parse v0.1.5
   Compiling duplicate v0.4.0
error[E0599]: no method named `get_ident` found for struct `syn::Path` in the current scope
   --> ~/.cargo/registry/src/index.crates.io-6f17d22bba15001f/derive-syn-parse-0.1.5/src/fields.rs:192:26
    |
192 |     let name = attr.path.get_ident()?.to_string();
    |                          ^^^^^^^^^ help: there is a method with a similar name: `is_ident`

error[E0599]: no method named `get_ident` found for struct `syn::Path` in the current scope
   --> ~/.cargo/registry/src/index.crates.io-6f17d22bba15001f/derive-syn-parse-0.1.5/src/variants.rs:154:26
    |
154 |     let name = attr.path.get_ident()?.to_string();
    |                          ^^^^^^^^^ help: there is a method with a similar name: `is_ident`

For more information about this error, try `rustc --explain E0599`.
error: could not compile `derive-syn-parse` (lib) due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
error: process didn't exit successfully: `~/.rustup/toolchains/stable-x86_64-apple-darwin/bin/cargo check --all-features --manifest-path ascent/Cargo.toml` (exit status: 101)
info: restoring <SNIP>/ascent/Cargo.toml
info: restoring <SNIP>/ascent_base/Cargo.toml
info: restoring <SNIP>/ascent_macro/Cargo.toml
error: process didn't exit successfully: `~/.rustup/toolchains/stable-x86_64-apple-darwin/bin/cargo hack check --workspace --all-features` (exit status: 1)
```

</details>

Bumping `syn` and `itertools` from …

```toml
[dependencies]
syn = { version = "1.0", features = [
    "derive",
    "full",
    "extra-traits",
    "visit-mut",
] }
itertools = "0.10"
# ...
```

… to …

```toml
[dependencies]
syn = { version = "1.0.109", features = [
    "derive",
    "full",
    "extra-traits",
    "visit-mut",
] }
itertools = "0.12"
# ...
```

fixes it.

<details>

<summary>Full terminal log</summary>

```terminal
$ cargo minimal-versions check --workspace --all-features --ignore-private -v                                                      (base) 
info: modifying from <SNIP>/ascent/ascent/Cargo.toml
info: modifying from <SNIP>/ascent/ascent_base/Cargo.toml
info: modifying from <SNIP>/ascent/ascent_macro/Cargo.toml
info: running `rustup run nightly cargo update -Z minimal-versions`
warning: virtual workspace defaulting to `resolver = "1"` despite one or more workspace members being on edition 2021 which implies `resolver = "2"`
note: to keep the current resolver, specify `workspace.resolver = "1"` in the workspace root's manifest
note: to use the edition 2021 resolver, specify `workspace.resolver = "2"` in the workspace root's manifest
note: for more details see https://doc.rust-lang.org/cargo/reference/resolver.html#resolver-versions
    Updating crates.io index
info: running `~/.rustup/toolchains/stable-x86_64-apple-darwin/bin/cargo hack check --workspace --all-features`
info: running `cargo check --all-features` on ascent (1/3)
warning: virtual workspace defaulting to `resolver = "1"` despite one or more workspace members being on edition 2021 which implies `resolver = "2"`
note: to keep the current resolver, specify `workspace.resolver = "1"` in the workspace root's manifest
note: to use the edition 2021 resolver, specify `workspace.resolver = "2"` in the workspace root's manifest
note: for more details see https://doc.rust-lang.org/cargo/reference/resolver.html#resolver-versions
    Blocking waiting for file lock on package cache
    Blocking waiting for file lock on package cache
    Blocking waiting for file lock on package cache
    Finished dev [unoptimized + debuginfo] target(s) in 0.68s

info: running `cargo check --all-features` on ascent_base (2/3)
warning: virtual workspace defaulting to `resolver = "1"` despite one or more workspace members being on edition 2021 which implies `resolver = "2"`
note: to keep the current resolver, specify `workspace.resolver = "1"` in the workspace root's manifest
note: to use the edition 2021 resolver, specify `workspace.resolver = "2"` in the workspace root's manifest
note: for more details see https://doc.rust-lang.org/cargo/reference/resolver.html#resolver-versions
    Finished dev [unoptimized + debuginfo] target(s) in 0.02s

info: running `cargo check --all-features` on ascent_macro (3/3)
warning: virtual workspace defaulting to `resolver = "1"` despite one or more workspace members being on edition 2021 which implies `resolver = "2"`
note: to keep the current resolver, specify `workspace.resolver = "1"` in the workspace root's manifest
note: to use the edition 2021 resolver, specify `workspace.resolver = "2"` in the workspace root's manifest
note: for more details see https://doc.rust-lang.org/cargo/reference/resolver.html#resolver-versions
    Finished dev [unoptimized + debuginfo] target(s) in 0.05s
info: restoring <SNIP>/ascent/ascent/Cargo.toml
info: restoring <SNIP>/ascent/ascent_base/Cargo.toml
info: restoring <SNIP>/ascent/ascent_macro/Cargo.toml
```

</details>

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -63,3 +63,23 @@ jobs:
     - name: Run macro tests
       working-directory: ./ascent_macro
       run: cargo test test  
+  
+  dependency-check:
+    runs-on: ubuntu-latest
+    
+    steps:
+      - uses: actions/checkout@v2
+      - run: cargo --version
+      
+      - name: Install Rust
+        uses: actions-rs/toolchain@v1
+        with:
+            override: true
+            toolchain: nightly
+
+      - uses: taiki-e/install-action@cargo-hack
+      - uses: taiki-e/install-action@cargo-minimal-versions
+
+      - name: cargo minimal-versions
+        run: cargo +stable minimal-versions check --workspace --all-features --ignore-private -v
+  

--- a/ascent_macro/Cargo.toml
+++ b/ascent_macro/Cargo.toml
@@ -13,11 +13,11 @@ proc-macro = true
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
-syn = { version = "1.0", features = ["derive", "full", "extra-traits", "visit-mut"] } 
+syn = { version = "1.0.109", features = ["derive", "full", "extra-traits", "visit-mut"] } 
 quote = "1.0"
 ascent_base = { version = "=0.5.0", path = "../ascent_base" }
 proc-macro2 = "1.0"
-itertools = "0.10"
+itertools = "0.12.0"
 petgraph = "0.6.0"
 derive-syn-parse = "0.1.5"
 lazy_static = "1.4.0"

